src/expr/parser.rs

```diff
@@ -1,58 +1,105 @@
-// Alex Ozdemir <aozdemir@hmc.edu> // <- Your name should replace this line!
+// Luis Viornery <lviornery@hmc.edu>
 // Starter code for HMC's MemorySafe, week 2
 //
 // The parser for an `Expr` (currently just produces the value fo the `Expr`)
 
+
+/*
+Workflow:
+start parsing additions => 
+start parsing multiplications => 
+start parsing literals => 
+finish parsing literals => 
+start parsing parens => 
+nested expr parsing => 
+finsish parsing parens => 
+finish parsing multiplication => 
+finish parsing additions
+*/
+
 use nom::digit;
 
 use std::str;
+use std::mem;
 use std::str::FromStr;
+use super::Expr;
+use super::BinOp;
 
-named!(parens<i64>, delimited!(
+//Make sure everything's return type is an expr
+
+named!(parens<Expr>,
+  delimited!(
     char!('('),
     expr,
     char!(')')
   )
 );
 
-named!(i64_literal<i64>,
-  map_res!(
+named!(literal_expr<Expr>,
+  map!(
     map_res!(
-      digit,
-      str::from_utf8
-    ),
-    FromStr::from_str
+        map_res!(
+          digit,
+          str::from_utf8
+        ),
+        FromStr::from_str
+      ),
+    //This wraps the isize in an expr literal
+    Expr::Literal
+  )
+);
+
+named!(literal_expr_wrapper<Expr>,
+  alt!(
+    //If we uncomment these, we get sad segfaults - this is a whitespace implementation attempt
+    //preceded!(char!(' '),literal_expr_wrapper) |
+    //terminated!(literal_expr_wrapper, char!(' ')) |
+    literal_expr
   )
 );
 
-named!(factor<i64>,
+named!(factor<Expr>,
   alt!(
-    i64_literal
+  literal_expr_wrapper
   | parens
   )
 );
 
 // we define acc as mutable to update its value whenever a new term is found
-named!(term <i64>,
+named!(term <Expr>,
   chain!(
     mut acc: factor  ~
              many0!(
                alt!(
-                 tap!(mul: preceded!(tag!("*"), factor) => acc = acc * mul) |
-                 tap!(div: preceded!(tag!("/"), factor) => acc = acc / div)
+                 //We have to map! to return a parser
+                 map!(preceded!(tag!("*"), factor), |next|{
+                 //mem::swap trickery to avoid moving acc into the closure
+                 let mut new_expr = Expr::Literal(0);
+                 new_expr = mem::replace(&mut acc, new_expr);
+                 mem::swap(&mut acc, &mut Expr::BinOp(Box::new(new_expr), BinOp::Times, Box::new(next)))}) |
+                 map!(preceded!(tag!("/"), factor), |next|{
+                 let mut new_expr = Expr::Literal(0);
+                 new_expr = mem::replace(&mut acc, new_expr);
+                 mem::swap(&mut acc, &mut Expr::BinOp(Box::new(new_expr), BinOp::Over, Box::new(next)))})
```

Comment:
Good use of `mem::replace`. However, since you write code like this 4 times, it'd be good to abstract it out.